Sources/BSON/BSONObjectID.swift

```diff
@@ -26,7 +29,22 @@ public struct BSONObjectID: Equatable, Hashable, CustomStringConvertible {
     ///   - `BSONError.InvalidArgumentError` if string passed is not a valid BSONObjectID
     /// - SeeAlso: https://github.com/mongodb/specifications/blob/master/source/objectid.rst
     public init(_ hex: String) throws {
-        fatalError("ah")
+        guard hex.utf8.count == (BSONObjectID.LENGTH * 2) else {
+            throw BSONError.InvalidArgumentError(message: "Cannot create ObjectId from \(hex)")
+        }
+        let len = hex.utf8.count / 2
+        var data = [UInt8](repeating: 0, count: 12)
+        for i in 0..<len {
+            let j = hex.index(hex.startIndex, offsetBy: i * 2)
+            let k = hex.index(j, offsetBy: 2)
+            let bytes = hex[j..<k]
+            if let num = UInt8(bytes, radix: 16) {
+                data[i] = num
+            } else {
+                throw BSONError.InvalidArgumentError(message: "Cannot create UInt8 from \(bytes)")
```

Comment:
i think it would be a bit clearer to say something like "Invalid hexadecimal character x" here